File: 14-weighted-sum-of-digits/task.ex
defmodule Task do

  def wsd(count, values) do
    wsd(count, values, [])
  end

  defp wsd(_, [], result) do
    Enum.reverse(result)
  end

  defp wsd(count, [head_value|tail_values], result) do
    size = String.length(Integer.to_string(head_value))
    res1 = w_sd(head_value, size)
    wsd(count - 1, tail_values, [Integer.to_string(res1)|result])
  end

  defp w_sd(a, s) do
    w_sd(a, s, 0)
  end

  defp w_sd(0, _, res) do
    res
  end
  
  defp w_sd(a, _, res) when a < 10 do
    res + a * 1
  end
  
  defp w_sd(a, s, res) do
    res1 = Kernel.rem(a, 10) * s
    w_sd(Kernel.div(a, 10), s - 1, res + res1)
  end

end

File: 14-weighted-sum-of-digits/task.erl
-module(task).

-export([wsd/2]).

wsd(Count, Values) ->
  wsd(Count, Values, []).

wsd(_, [], Result) ->
  lists:reverse(Result);
wsd(Count, [ValueHead|ValuesTail], Result) ->
  Res1 = w_sd(ValueHead, erlang:byte_size(erlang:integer_to_binary(ValueHead))),
  wsd(Count - 1, ValuesTail, [erlang:integer_to_binary(Res1)|Result]).

w_sd(A, S) ->
  w_sd(A, S, 0).

w_sd(0, _, Res) ->
  Res;
w_sd(A, _, Res) when A < 10 ->
  Res + A * 1;
w_sd(A, S, Res) ->
  w_sd(A div 10, S - 1, Res + (A rem 10) * S).
